src/Microsoft.Graph.Cli.Core/IO/TableOutputFormatter.cs

```diff
@@ -0,0 +1,98 @@
+            var table = this.ConstructTable(doc);
+            AnsiConsole.Write(table);
+        }
+
+        private Table ConstructTable(JsonDocument document) {
+            var root = document.RootElement;
+            JsonElement firstElement;
+            JsonElement value;
+            var hasValueProperty = root.TryGetProperty("value", out value);
+            if (hasValueProperty) {
+                root = value;
+            }
+
+            if (root.ValueKind == JsonValueKind.Array && root.GetArrayLength() > 0) {
+                var enumerated = root.EnumerateArray();
+                firstElement = enumerated.FirstOrDefault();
+            } else {
+                firstElement = root;
+            }
+
+            var properties = new List<(JsonValueKind,string)>();
+            var table = new Table().Expand();
+            if (firstElement.ValueKind != JsonValueKind.Object) {
+                properties.Add((firstElement.ValueKind, "Value"));
+            } else {
+                var restrictedValueKinds = new JsonValueKind[] {
+                    JsonValueKind.Array,
+                    JsonValueKind.Object
+                };
+                properties = firstElement.EnumerateObject()
+                    .Where(p => !restrictedValueKinds.Contains(p.Value.ValueKind))
+                    .Select(p => (p.Value.ValueKind, p.Name)).ToList();
+            }
+
+            if (root.ValueKind == JsonValueKind.Array) {
+                foreach (var property in properties)
+                    table.AddColumn(property.Item2, column => {
+                        if (property.Item1 == JsonValueKind.Number)
+                            column.RightAligned().PadLeft(10);
+                    });
+
+                foreach (var row in root.EnumerateArray())
+                {
+                    var rowCols = properties.Select(p => {
+                        var valueKind = p.Item1;
+                        var propertyName = p.Item2;
+                        JsonElement property;
+                        var hasProp = row.TryGetProperty(propertyName, out property);
+                        if (hasProp) {
+                            object? value = null;
+                            switch (valueKind)
+                            {
+                                case JsonValueKind.String:
+                                    value = property.GetString();
+                                    break;
+                                case JsonValueKind.True:
+                                case JsonValueKind.False:
+                                    value = property.GetBoolean();
+                                    break;
+                                case JsonValueKind.Number:
+                                    value = property.GetDecimal();
+                                    break;
+                            }
+                            return new Markup(value?.ToString() ?? "-");
```

Comment:
If auto-truncation is available, maybe we can just print the backing JSON text directly, that has the ability to show an array value as well. For the types that are supported here, the output stays almost the same with `GetRawText()`:
- for `string`, quotation marks are added (i.e. `value` vs `"value"`)
- boolean values are normalized to JSON representation (e.g. `true`) instead of string representation of a boolean in .NET (e.g. `True`).

What do you think?
```suggestion
                            return new Markup(property.GetRawText());
```

Reply:
Thanks for the suggestion. I like that it's less code for a similar result. However, I do have a few concerns on the experience that I'd like to address with the product team at the moment. For example, if a property is of type `Array`, it could be an array of objects and at a certain point, the truncated value would not be useful to the users which ends up taking space for no added value. For object arrays, the property names will be printed in the cell and take up space. I can see a situation where a column would only show `[{"propertyName...` as the value in the cell. We'd talked about selecting default properties to show initially and allow users to override that. Another concern I have is on the added logic that we'll ultimately need to sanitize the table output. e.g. removing the string `null` from the output for null values, formatting dates and numeric outputs, handling empty arrays and objects etc.

My plan was to add the display of nested object values first and prefix the column name with the property name that the object appears in. For example, given the json object below:

```json
{
    "a": "value a",
    "b": {
        "c": 10,
        "d": "value d"
    }
}
```
The output table would have the output like below:
```shell
---------------------------------
|  a        |  b.c  |  b.d      |
---------------------------------
|  value a  |  10   |  value d  |

```

This would allow for a better experience if we want to support property selection by the user. So they can decide they only want to see properties `a` and `a.d`. It also means the cells in the table aren't cluttered with unnecessary information like property names since there's limited real estate already.
I haven't decided what to do about arrays yet, but I'm open to ideas

Reply:
Thanks for the detailed response. It all makes sense to me. A small addition: PowerShell truncates nested objects with an overview of what is nested inside. That output is sometimes useful, sometimes not, especially when the overview contains only keys and no values. But at least it gives a clue on how to expand the object. If we can find design documents for PowerShell's table formatter, we may find additional requirements, tradeoffs and insights there. Obviously we are not writing a full shell here,so we don't have to support all of their requirements :).